core/ir/x86/decode.c

```diff
@@ -2753,6 +2755,15 @@ byte *
 decode(void *drcontext, byte *pc, instr_t *instr)
 {
     dcontext_t *dcontext = (dcontext_t *)drcontext;
+
+    /* Synthetic ISA has its own decoder.
+     * XXX i#1684: when DR can be built with full dynamic architecture selection we won't
+     * need to pollute the decoding of other architectures with this synthetic ISA special
+     * case.
+     */
+    if (instr_get_isa_mode(instr) == DR_ISA_SYNTHETIC)
+        return decode_from_synth(dcontext, pc, instr);
```

Comment:
decode_common() is called by other functions, hence we move decode_from_synth() from decode_common()
to decode().